"# From raw reads to processed alignments\n\nInstructor: **Cinta Pegueroles Queralt**\n\nThe goal of this practical session is to learn how to obtain a <em>de novo</em> transcriptome from RNAseq data. The practical is organized in four main steps:\n\n1. **Sanitation**: pre-processing RNAseq data  \n2. **<em>de novo</em> assembly**: in the absence of a reference genome\n3. **QC**: assessing the quality of de novo transcriptome\n4. **Post-processing analyses**\n\n\n### 0. Downloading data:\nDue to time and computer resources we will work with a test data set from [TRINITY GITHUB][https://github.com/trinityrnaseq/trinityrnaseq/tree/devel/sample_data/test_Trinity_Assembly]\n\n```\ncd session2\nmkdir 0data\ncd 0data\nwget https://github.com/trinityrnaseq/trinityrnaseq/raw/devel/sample_data/test_Trinity_Assembly/reads.left.fq.gz\nwget https://github.com/trinityrnaseq/trinityrnaseq/raw/devel/sample_data/test_Trinity_Assembly/reads.right.fq.gz\n```\n\n### 1.1. Quality control of RNAseq data\nNow we will perform the quality control (QC) of the RNAseq data using the software FastQC. First of all, we will create a folder for the QC output files\n\n```\ncd session2\nmkdir 1QC\n```\nFirst we will create/edit a bash script to run FastQC software (fastqc.sh). IMPORTANT: mind to replace xxx by the current path.\n\n```\n#!/bin/bash\n\n# define names\n#SBATCH --job-name=FastQC\n#SBATCH --error FastQC-%j.err\n#SBATCH --output FastQC-%j.out\n\n# memory and CPUs request\n#SBATCH --mem=10G\n#SBATCH --cpus-per-task=5\n\n# module load\nmodule load fastqc/0.11.9\n\n# jobs to launch\nsrun fastqc -t 5 /xxx/session2/0data/reads.left.fq.gz -o /xxx/session2/1QC\nsrun fastqc -t 5 /xxx/session2/0data/reads.right.fq.gz -o /xxx/session2/1QC\n```\nTo launch FastQC script in the cluster write on the terminal:\n```\nsbatch fastqc.sh\n```\nUse \"s_jobs\" command to check the state of the process. To check the results you need to download the files on a local computer. WARNING: mind to change the paths accordingly\n```\nscp -r user@cluster:/my_CLUSTER_path_to_session2/* my_LOCAL_path_to_session2/\n```\n\nAlternatively we could run the FastQC software directly on the terminal:\n```\ncd session2\nfastqc -o 1QC 0data/reads.left.fq.gz 0data/reads.right.fq.gz\n```\n\n### 1.2. Trimming sequences\n\nIf we detect a drop in the quality of the sequences and/or the presence of adapters we need to trim our sequences. We will do it by using the Trimmomatic software.\n\nFirst, we will create a folder for the trimmed sequences:\n\n```\ncd session2\nmkdir 2trimmed_data\ncd 2trimmed_data\n\n```\nThen we will create/edit a bash script to run Trimmomatic software (trimseq.sh). IMPORTANT: mind to replace xxx by the current path.\n\n```\n#!/bin/bash\n\n# define names\n#SBATCH --job-name=trimmomatic\n#SBATCH --error trimmomatic-%j.err\n#SBATCH --output trimmomatic-%j.out\n\n# memory and CPUs request\n#SBATCH --mem=10G\n#SBATCH --cpus-per-task=5\n\n# module load\nmodule load java\nmodule load trimmomatic\n\n# jobs to launch\nsrun java -jar /cluster/software/Trimmomatic/Trimmomatic-0.39/trimmomatic-0.39.jar PE -threads 5 \\\n        /xxx/session2/0data/reads.left.fq.gz \\ \n        /xxx/session2/0data/reads.right.fq.gz \\\n        /xxx/session2/2trimmed_data/reads_1.P.fq.gz \\ \n        /xxx/session2/2trimmed_data/reads_1.U.fq.gz \\\n        /xxx/session2/2trimmed_data/reads_2.P.fq.gz \\\n        /xxx/session2/2trimmed_data/reads_2.U.fq.gz \\\n        ILLUMINACLIP:/xxx/session2/TruSeq3-PE.fa:2:30:1 \\\n        LEADING:3 TRAILING:3 SLIDINGWINDOW:4:15 MINLEN:36\n```\n\nTo launch the Trimmomatic script in the cluster write on the terminal:\n```\nsbatch trimSeq.sh\n```\nUse \"s_jobs\" command to check the state of the process. \n\nAlternatively we could run the Trimmomatic software directly on the terminal:\n\n```\ntrimmomatic PE ../0data/reads.left.fq.gz ../0data/reads.right.fq.gz \\\n    ../2trimmed_data/reads_1.P.fq.gz ../2trimmed_data/reads_1.U.fq.gz \\\n    ../2trimmed_data/reads_2.P.fq.gz ../2trimmed_data/reads_2.U.fq.gz \\\n    ILLUMINACLIP:TruSeq3-PE.fa:2:30:1 LEADING:3 TRAILING:3 SLIDINGWINDOW:4:15 MINLEN:36\n```\n\n### 1.3. Quality control of the trimmed sequences\nNow we will perform the quality control (QC) of the trimmed sequences using again the software FastQC. First of all, we will create a folder for the QC output files\n\n```\ncd session2\nmkdir 3QC_trimmed\n```\nSince now we have 4 files to check (2 paired, 2 unpaired) we will use a loop to avoid repeating commands. We will create/edit a bash script to run FastQC software in loop over all trimmed sequences (fastqc_loop.sh). IMPORTANT: mind to replace xxx by the current path.\n\n```\n#!/bin/bash\n\n# define names\n#SBATCH --job-name=FastQC_loop\n#SBATCH --error FastQC_loop-%j.err\n#SBATCH --output FastQC_loop-%j.out\n\n# memory and CPUs request\n#SBATCH --mem=10G\n#SBATCH --cpus-per-task=5\n\n# module load\nmodule load fastqc/0.11.9\n\n# jobs to launch\nfor file in /xxx/session2/2trimmed_data/*fq.gz;\n        do\n                srun fastqc -t 5 $file -o /xxx/session2/3QC_trimmed\n        done\n```\n\nTo launch fastQC_loop.sh script in the cluster write on the terminal:\n```\nsbatch fastqc_loop.sh\n```\nUse \"s_jobs\" command to check the state of the process. To check the results you need to download the files on a local computer. WARNING: mind to change the paths accordingly\n```\nscp -r user@cluster:/my_CLUSTER_path_to_session2/* my_LOCAL_path_to_session2/\n```\nAlternatively we could run the FastQC software directly on the terminal:\n```\ncd session2\nfor file in /xxx/session2/2trimmed_data/*fq.gz; do fastqc $file -o /xxx/session2/3QC_trimmed; done\n```\n\n### 2. <em>De novo</em> transcriptome assembly\n### 2.1. Running trinity software\n\nNow we will obtain a <em>de novo</em> transcriptome using the trimmed paired end sequences. In the same script we will also calculate some stats on the obtained transcriptome. First, we will create two folders to store the analysis:\n\n```\ncd session2\nmkdir 4trinity #will contain the de novo transcriptome\n```\n\nFirst, we will create/edit a bash script to run trinity software (trinity.sh) to obtain a <em>de novo</em> transcriptome. IMPORTANT: mind to replace xxx by the current path.\n```\n#!/bin/bash\n\n# define names\n#SBATCH --job-name=trinity\n#SBATCH --error trinity-%j.err\n#SBATCH --output trinity-%j.out\n\n# memory and CPUs request\n#SBATCH --mem=100G\n#SBATCH --cpus-per-task=10\n\n# module load\nmodule load trinity\n\n# jobs to launch\nsrun Trinity --left /xxx/session2/2trimmed_data/reads_1.P.fq.gz \\\n\t--right /xxx/session2/2trimmed_data/reads_2.P.fq.gz --seqType fq \\\n\t--normalize_reads --normalize_max_read_cov 30 \\\n\t--max_memory 100G \\\n\t--CPU 10 \\\n\t--output /xxx/session2/4trinity\n\n```\nTo launch the trinity script in the cluster write on the terminal:\n```\nsbatch trinity.sh\n```\nUse \"s_jobs\" command to check the state of the process. \n\nAlternatively we could run the Trinity software directly on the terminal:\n```\ncd session2/4trinity\n\n#launch trinity\nTrinity --left /xxx/session2/2trimmed_data/reads_1.P.fq.gz \\\n\t--right /xxx/session2/2trimmed_data/reads_2.P.fq.gz --seqType fq \\\n\t--normalize_reads --normalize_max_read_cov 30 \\\n\t--max_memory 100G \\\n\t--CPU 10 \\\n\t--output /cluster/home/cpegueroles/phylo/session2/4trinity\n```\n### 2.1. cd-hit: redundance reduction\n\nNow we will avoid reduncamce in the de novo transcriptome by using the cdhit software. First we will create a folder to drop the analysis:\n\n```\ncd session2\nmkdir 5cdhit\n```\n\nFirst we will create/edit a bash script to run cdhit software (cdhit.sh). IMPORTANT: mind to replace xxx by the current path.\n```\n#!/bin/bash\n\n# define names\n#SBATCH --job-name=cdhit\n#SBATCH --error cdhit-%j.err\n#SBATCH --output cdhit-%j.out\n\n# memory and CPUs request\n#SBATCH --mem=50G\n#SBATCH --cpus-per-task=5\n\n# module load\nmodule load cdhit\n\ncd-hit-est -i /xxx/session2/4trinity/Trinity.fasta -o /xxx/session2/5cdhit/Trinity_cdhit.fasta -c 0.9 -M 50 -T 5 > /xxx/session2/5cdhit/Trinity_cdhit.err\n\n```\nTo launch the script in the cluster write on the terminal:\n```\nsbatch cdhit.sh\n```\nAlternatively we could run the cdhit software directly on the terminal:\n```\ncd session2/5cdhit\n\ncd-hit-est -i /xxx/session2/4trinity/Trinity.fasta -o /xxx/session2/5cdhit/Trinity_cdhit.fasta -c 0.9 -n 8 -M 0 -T 0 -d 40 > /xxx/session2/5cdhit/Trinity_cdhit.err\n\n\n```\n### 3. Post-processing\n\nIn this example we are analyszing the de novio transcriptoem obtained with trinity, but we could apply the same analysis to the reduced fasta file obtained after running cdhit software\n\n### 3.1. Basic stats\nWe will check basic statistics of both trinity assembly and CDhit reduced transcriptome\n\n```\ncd session2\nmkdir 6QC_trinity \n```\n\nFirst we will create/edit a bash script to run the stats (trinityQC.sh) to obtain a de novo transcriptome. IMPORTANT: mind to replace xxx by the current path.\n\n```\n#!/bin/bash\n\n# define names\n#SBATCH --job-name=trinityQC\n#SBATCH --error trinityQC-%j.err\n#SBATCH --output trinityQC-%j.out\n\n# memory and CPUs request\n#SBATCH --mem=10G\n#SBATCH --cpus-per-task=5\n\n# module load\nmodule load trinity\n\n# jobs to launch\n\nsrun /cluster/software/trinity/trinityrnaseq-v2.11.0/util/TrinityStats.pl /xxx/session2/4trinity/Trinity.fasta > /xxx/session2/6QC_trinity/Trinity_assembly.metrics\nsrun /cluster/software/trinity/trinityrnaseq-v2.11.0/util/TrinityStats.pl /xxx/session2/5cdhit/Trinity_cdhit.fasta > /xxx/session2/6QC_trinity/Trinity_cdhit_assembly.metrics\n```\nTo launch the script in the cluster write on the terminal:\n```\nsbatch trinityQC.sh\n```\n\n### 3.2. Representation of reads\nIn this step we will backmap the RNAseq reads to the de novo assembly to check the completeness of the transcriptome. The higher the precentge of reads that backmap, the better.\n\nTo do so, we will use the paired trimmed reads (which are those that we used for the de novo assembly) and the software hisat2 for mapping.\n\nFirst we will create/edit a bash script to run hisat2 software (hisat2.sh). IMPORTANT: mind to replace xxx by the current path.\n\n```\n#!/bin/bash\n\n# define names\n#SBATCH --job-name=hisat2\n#SBATCH --error hisat2-%j.err\n#SBATCH --output hisat2-%j.out\n\n# memory and CPUs request\n#SBATCH --mem=50G\n#SBATCH --cpus-per-task=5\n\n# module load\nmodule load hisat/2.2.1\n\n# jobs to launch\nsrun hisat2-build /xxx/session2/trinity/Trinity.fasta /xxx/session2/4trinity/Trinity\n\nsrun hisat2 -p 10 -x /xxx/session2/4trinity/Trinity \\\n\t-1 /xxx/session2/2trimmed_data/reads_1.P.fq.gz -2 /xxx/session2/2trimmed_data/reads_2.P.fq.gz \\\n\t-S /xxx/session2/6QC_trinity/reads.sam &> /xxx/session2/6QC_trinity/reads.sam.info\n```\n\n### 3.3. From ncl to aa: TransDECODER\n\nWe will use transdecoder to get our predicted proteome. First, we will create a folder to drop the analysis, we will move to this folder and we will launch the analysis.\n\nFirst we will create/edit a bash script to run transdecoder software (transdecoder.sh). IMPORTANT: mind to replace xxx by the current path.\n\n```\ncd session2\nmkdir 6proteome\ncd  6proteome\n```\n```\n#!/bin/bash\n\n# define names\n#SBATCH --job-name=transdecoder\n#SBATCH --error transdecoder-%j.err\n#SBATCH --output transdecoder-%j.out\n\n# memory and CPUs request\n#SBATCH --mem=10G\n#SBATCH --cpus-per-task=10\n\n# module load\nmodule load transdecoder\n\n# jobs to launch\n\nsrun TransDecoder.LongOrfs -t /xxx/session2/4trinity/Trinity.fasta \n```\n\nTo launch the script in the cluster write on the terminal:\n```\nsbatch transdecoder.sh\n```\n\n#### Bibliography\n\nGrabherr MG, Haas BJ, Yassour M, Levin JZ, Thompson DA, Amit I, Adiconis X, Fan L, Raychowdhury R, Zeng Q, Chen Z, Mauceli E, Hacohen N, Gnirke A, Rhind N, di Palma F, Birren BW, Nusbaum C, Lindblad-Toh K, Friedman N, Regev A. Full-length transcriptome assembly from RNA-seq data without a reference genome. Nat Biotechnol. 2011 May 15;29(7):644-52. doi: 10.1038/nbt.1883. PubMed PMID: 21572440.\n\n\nHaas BJ, Papanicolaou A, Yassour M, Grabherr M, Blood PD, Bowden J, Couger MB, Eccles D, Li B, Lieber M, Macmanes MD, Ott M, Orvis J, Pochet N, Strozzi F, Weeks N, Westerman R, William T, Dewey CN, Henschel R, Leduc RD, Friedman N, Regev A. De novo transcript sequence reconstruction from RNA-seq using the Trinity platform for reference generation and analysis. Nat Protoc. 2013 Aug;8(8):1494-512. Open Access in PMC doi: 10.1038/nprot.2013.084. Epub 2013 Jul 11. PubMed PMID:23845962."